### DISCUSSION TITLE
How to render text-paths as HATCH entities

### DISCUSSION THREAD

**Discussion by shine-flower** [mozman/ezdxf#742] [Q&A] [answered]
Implement similar functions #715 

> The following code gets some paths, how to fill the lines in the text, you can set the density of the kind, fill a line every 1 mm.
```code
doc = ezdxf.readfile('211.dxf')
msp = doc.modelspace()
height = 2.7
first_path = text2path.make_paths_from_str("厂牌AAA123", fonts.FontFace(family="SimHei"), (height * 1.45))
final_paths = path.transform_paths(first_path, Matrix44.translate(1.810846223812833, 1.482741896976211, 0))

path.render_splines_and_polylines(msp, final_paths, dxfattribs={"color": 1})
zoom.extents(msp)
doc.saveas("text2path.dxf")
```
> Achieve the following results, Is there any demo or method to implement this function.
![image](https://user-images.githubusercontent.com/105044629/190543675-fb1640f6-e4d1-4546-97a2-075084d5e985.png)

**Comment by mozman** [comment, marked as answer]
You have to render the text as HATCH entities. The first example shows how to render HATCHES with solid filling:

![image](https://user-images.githubusercontent.com/4230966/190837617-7f715a38-1ea0-4f9d-8521-a6e3eb87413d.png)

```python
import ezdxf
from ezdxf.addons import text2path
from ezdxf.tools import fonts
from ezdxf.math import Matrix44
from ezdxf import zoom

doc = ezdxf.new()
msp = doc.modelspace()

hatches = text2path.make_hatches_from_str(
    "厂牌AAA123",
    font=fonts.FontFace(family="SimSun"),
    size=4,
)
m = Matrix44.translate(2, 1.5, 0)

for hatch in hatches:
    hatch.transform(m)
    hatch.set_solid_fill(color=1, style=0)
    msp.add_entity(hatch)

zoom.extents(msp)
doc.saveas("text2solid_hatch.dxf")
```

The `HATCH` entity can be filled with predefined patterns, but this involves much guessing:

![image](https://user-images.githubusercontent.com/4230966/190838468-010f4544-9d7a-46a0-8711-e510dba4d194.png)

```python
...

for hatch in hatches:
    hatch.transform(m)
    hatch.set_pattern_fill(name="ANSI31", color=1, scale=0.02, angle=-45, style=0)
    msp.add_entity(hatch)

...
```

You have more control over the HATCH pattern by defining you own custom pattern:

![image](https://user-images.githubusercontent.com/4230966/190838420-1d366f4e-b64e-4851-a55e-096f805e00bb.png)

```python
...

# ANSI31 definition: [angle, base_point, offset, dash_length_items]
# "ANSI31": [[45.0, (0.0, 0.0), (-2.2450640303, 2.2450640303), []]]
OFFSET = (0, 0.05)  # (x, y)
SOLID = []  # continuous line
DASHED = [0.5, -0.3]  # > 0 line, < 0 gap, 0 = point
MY_PATTERN = [
    [0.0, (0.0, 0.0), OFFSET, SOLID],  # a single hatch line definition
]

for hatch in hatches:
    hatch.transform(m)
    hatch.set_pattern_fill(
        name="MY_PATTERN",
        color=1,
        style=0,
        pattern_type=2,
        definition=MY_PATTERN,
    )
    msp.add_entity(hatch)

...
```

For adding the outline path you have to render the text a second time as paths as in your original script.

**Reply by mozman** [reply]
Added an example script for this topic: https://github.com/mozman/ezdxf/blob/master/examples/addons/text_string_to_hatch.py

**Reply by shine-flower** [reply]
> 为此主题添加了示例脚本：[https ://github.com/mozman/ezdxf/blob/master/examples/addons/text_string_to_hatch.py](https://github.com/mozman/ezdxf/blob/master/examples/addons/text_string_to_hatch.py)

Thank you very much for your reply.

**Reply by shine-flower** [reply]
> After the above operation is completed, the filling is realized, which is a block in DXF, and it can not recognize the concept of block in the laser marking machine. I'll go into AutoCAD and run the 'explode' command, and the laser marker will identify the filled line.Does EZDXF have a way to implement the AutoCAD 'explode' function?
Before use 'explode' command
![image](https://user-images.githubusercontent.com/105044629/190960476-9bc55c02-4816-4478-b583-de4d099a2ad7.png)
After use 'explode' command
![image](https://user-images.githubusercontent.com/105044629/190960314-cf73f038-bf05-4a08-ade5-501ab1d4f58f.png)
